### PR TITLE
BackgroundTaskManager

### PR DESCRIPTION
**Pull request by UselessToucan**
I guess this is enough to start a discussion of #1097 
1. Is this approach acceptable? I mean, store background tasks inside of DI cached object and use this object for various management activities(start new task, cancel existing task and so on).
2. Minimal set of requirements for background tasks management?
I mean something like this:
- [ ] Create background task from Action
- [ ] Create background task from Task object
- [ ] Cancel task which is in pending or running state
- [ ] Continuation support
- [ ] Provide task description which can be shown to the end user
3. What to do with task cancellation? TPL has built-in cancellation support, but osu! also has it's own cancellation inside of ProgressNotification. At the moment, the second approach is used for beatmap import.
4. How to handle closing the game when there are active background tasks? Is it ok to display list of active tasks with "force close" button?

### CODE REVIEW

File: osu.Framework/Threading/BackgroundTaskManager.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+namespace osu.Framework.Threading
+{
+    /// <summary>
+    /// Encapsulates tasks which were offloaded to background threads
+    /// </summary>
+    public class BackgroundTaskManager
+    {
+        private static readonly object instance_lock = new object();
+        private static BackgroundTaskManager instance;
```
Comment (Tom94) [inline osu.Framework/Threading/BackgroundTaskManager.cs]: I'd prefer if this class did not have any static members and that the singleton would then reside in a higher-up class like `Game` as e.g. `Game.BackgroundTaskManager`. Classes themselves should make as few assumptions about how they're going to be used as possible.

Comment (smoogipoo) [reply osu.Framework/Threading/BackgroundTaskManager.cs]: Could probably be DI'd too

Comment (Tom94) [reply osu.Framework/Threading/BackgroundTaskManager.cs]: I'm all for DI if that means we can avoid static. :)

**Comment by UselessToucan** [comment]
I will try to update this PR on weekend. I'm planning to introduce separate class for representing background task and do something more valuable than just storing Task objects in the List.

**Comment by smoogipoo** [comment]
My stance on this right now is that it isn't really useful in its current implementation. We essentially need two use cases:

1. Concurrency limiting. This can be achieved by deriving TaskScheduler.
2. Global task cancellation. This can be achieved by providing a factory similar to what you have here, which provides a `StartNew` function.

In the case of 2., the factory should contain linked cancellation sources between a cancellation token provided by the user to `StartNew` and a local cancellation token. Exposing a `CancelAll` method should basically trigger a cancel on the linked sources and then `WaitAll` on all the running tasks.

Essentially `StartNew` should be `Task StartNew(Action<CancellationToken> token, CancellationToken cancellationToken)`.

**Comment by UselessToucan** [comment]
@smoogipoo In general, I agree with you. I'll try to find some time to rewrite this.
I would like to clarify how should cancellation work for individual task. It can be done by just calling `CancellationTokenSource.Cancel` method from the outside. Or I guess I can hide this Token/TokenSource stuff inside of the task manager and provide a couple of public methods like `Cancel(int taskId)`.

**Comment by smoogipoo** [comment]
Consider importing from stable. There are two forms of cancellation we want:
1. A cancellation handled by, probably, the notification.
2. A cancellation handled by the game upon exiting (or something like that).

So individual classes will handle the creation, management, and cancellation of the tasks locally, through cancellation tokens, as is currently done with tasks. E.g. clicking the X button in a notification should cancel on the specific cancellation token.

At the same time, the game can call myFactory.CancelAll() to cancel all tasks, which in-turn uses the cancellation tokens local to the factory.

**Comment by smoogipoo** [comment]
I don't believe the factory should provide a singular way to Cancel a task, but that should be easily doable in case we need it - just single out the relevant cancellation source local to the factory and call Cancel on it.

**Comment by smoogipoo** [comment]
Closing for now, re-open when re-implemented as described. Feel free to poke on discord to discuss if needed.

**Pull request closed without merging**